tools/codegen/gen.py

```diff
@@ -837,9 +847,55 @@ def write_with_template(self, filename: str, template_fn: str,
                 assert_never(env)
 
 
-    def write(self, filename: str, env_callable: Callable[[], Union[str, Union[str, Dict[str, object]]]]) -> None:
+    def write(self, filename: str, env_callable: Callable[[], Union[str, Union[str, Dict[str, Any]]]]) -> None:
         self.write_with_template(filename, filename, env_callable)
 
+    def write_sharded(
+            self,
+            filename: str,
+            items: List[T],
+            *,
+            key_fn: Callable[[T], str],
+            env_callable: Callable[[T], Dict[str, List[str]]],
+            num_shards: int,
+            base_env: Dict[str, Any] = {},
+    ) -> None:
+        everything: Dict[str, Any] = {**base_env, 'shard_id': 'Everything'}
+        shards: List[Dict[str, Any]] = [
+            {**base_env, 'shard_id': f'_{i}'} for i in range(num_shards)]
+
+        def merge_env(into: Dict[str, List[str]], from_: Dict[str, List[str]]):
+            for k, v in from_.items():
+                if k in into:
+                    assert isinstance(into[k], list)
+                    into[k] += v
+                else:
+                    into[k] = list(v)
```

Comment:
isn't v already a list here?

Reply:
That's what I thought as well. But if you don't copy the list here, then the same list will exist in two dicts and items will get re-appended into the same list.

Reply:
Ok!